src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -372,7 +372,106 @@ protected void InitializeReferences()
             RegexRunner runner = Interlocked.Exchange(ref _runner, null) ?? CreateRunner();
             try
             {
-                return runner.Scan(this, input, beginning, beginning + length, startat, prevlen, quick, internalMatchTimeout);
+                runner.InitializeTimeout(internalMatchTimeout);
+                runner.runtext = input;
+                ReadOnlySpan<char> span = input.AsSpan(beginning, length);
+                runner.InitializeForScan(this, span, startat - beginning, quick);
+
+                int stoppos = RightToLeft ? 0 : span.Length;
+
+                // If previous match was empty or failed, advance by one before matching.
+                if (prevlen == 0)
+                {
+                    if (runner.runtextstart == stoppos)
+                    {
+                        return RegularExpressions.Match.Empty;
+                    }
+
+                    runner.runtextpos += RightToLeft ? -1 : 1;
+                }
+
+                return InternalPerformScan(quick, input, beginning, runner, span, returnNullIfQuick: true);
+            }
+            finally
+            {
+                _runner = runner;
+            }
+        }
+
+        private static Match? InternalPerformScan(bool quick, string input, int beginning, RegexRunner runner, ReadOnlySpan<char> span, bool returnNullIfQuick)
+        {
+            // We need to save the original beginning of the input in case we are in the CompileToAssembly case, which relies on
+            // setting runtextbeg and runtextend correctly. We cannot calculate these values on the fly based on input and the span
+            // because it is not guaranteed that span.Length + beginning = input.
+            runner.originalRuntextbeg = beginning;
+            runner.Scan(span);
+
+            Match? match = runner.runmatch;
+            // if we got a match, set runmatch to null if quick is true
+            if (match!.FoundMatch)
+            {
+                runner.runtext = null; // drop reference to text to avoid keeping it alive in a cache
+
+                if (match.Text != input)
+                {
+                    match.Text = input;
+                }
```

Comment:
What's the purpose of this?  I'm still confused by the logic here only being partially implemented.  In what situation won't they be the same?  And if this is needed, shouldn't it be below the subsequent if block?  Otherwise we're setting match.Text only to set it to null immediately after.

Reply:
Before my PR, when we called InitializeForGo() we would initialize runmatch to have the Text field set to the value of runtext. Since we are no longer setting runtext, we had to move the setting of that Text field to Run once that runmatch object is prepared to be returned to the user. One thing we could do here is to also condition this block on whether quick is true or not, since if it is true, we won't be handing the match object to the user anyway.

Reply:
Ok, so then in what situation does match.Text == input?  I'm wondering why this isn't just `match.Text = input;` _after_ the subsequent if block.

Reply:
That is a good point, I think that is never the case but I will double check. IIRC I added the ifcheck in response to a comment you gave (on a different line of code) that we should only set the text when it is not already the same value, but I think at the time I misunderstood that comment of yours and applied it here incorrectly. I will remove the ifcheck after validating this is never the case.